# 第二章 整洁代码的模式

## 2.1 使用断言
有时候，一些真正有用的语言特性往往没有得到应有的重视。Python的`assert`语句就是如此

在这一章我将会介绍如何使用Python的断言。你会学到如何使用断言自动捕获代码中的错误，
这会增加代码的可读性，会让你的代码更加容易调试。

在这个点上，你可能会疑惑，"什么是断言，断言有什么好处？"

Python的断言可以通过测试一个条件辅助调试。如果断言的条件是true，什么都不会发生，代码继续正常执行。
但是如果条件是false，会抛出一个`AssertionError`的异常，这个异常可以添加一个可选的错误信息


### Python中断言的一个例子
在这个简单的例子中，你可以看到断言的作用。我尽量使这个例子和真实世界中你可能遇到的问题更相似

假设你在使用Python搭建一个在线商店。你写下了下面的apply_discount函数为系统添加一个打折优惠的功能。

    def apply_discount(product, discount):

         price = int(product['price'] * (1.0 - discount))

         assert 0 <= price <= product['price']

         return price


上面函数中的assert语句会保证在任何情况下，经过这个函数计算后的打折后的价格都不会比0更低，不会比初始价格更高。

让我们确认一下，在我们调用这个函数计算一个折扣的时候，这个逻辑确实按照我们设想的方式工作。
在这个例子中，我们商店中的商品会使用字典进行表示。这很可能和你在真实应用中的做法不同，
但是对于演示断言，这已经足够了。我们先创建一个商品当做例子，这个商品是一双价值$149的漂亮鞋子

    shoes = {'name': 'Fancy Shoes', 'price': 14900}

顺便说一下，你有没有注意到我使用整数来表示以分为单位的价格，从而避免了找零时的四舍五入问题？
这是一个好主意，但是我不这么认为。
现在，如果给予25%的折扣，销售价格就会是$111.75

    apply_discount(shoes, 0.25)
    >>> 11175

一切正常。现在我们试试一些无效的折扣，比如200%的折扣，将会让我们倒付钱给消费者
    >>> apply_discount(shoes, 2。0)

    Traceback (most recent call last):
    
    File "<input>", line 1, in <module>
    
    apply_discount(prod, 2.0)
    
    File "<input>", line 4, in apply_discount
    
    assert 0 <= price <= product['price']
    
    AssertionError

你看，当我门尝试打一个无效的折扣的时候，我们的程序抛出了一个`AssertionError`。
之所以会这样是因为200%的折扣违反了我们在`apply_discount`函数中的断言条件
你还能看到异常错误栈指出了包含失败断言的哪一行代码，
如果你(或者你们团队的其它开发者)在测试在线商店的时候遇到了这样的一个错误，
通过错误栈会很容易找到具体发生了什么

这会在很大成都上加快调试的速度，并且会让你的代码更易于维护。这就是断言的力量


### 为什么不直接使用一个常规的异常
你可能会疑惑为什么在上面的例子中我不直接使用`if`语句抛出一个异常。

正确使用断言可以通知开发者那些程序中不可恢复的错误。断言不是为了提示预期中的错误，
对于像文件没找到这种预期中的错误，用户可以采取纠正措施，或者直接重试

断言意味着程序的内部自检。断言会声明一些不可能发生的条件，如果这些条件没有被兜住，
意味着代码中存在bug

如果你的代码没有任何bug，这些断言条件永远不会发生。但是一旦发生，程序就会崩溃，此时会抛出一个断言错误，
告诉你哪个"不可能"的条件被触发了。
这会让定位和修复bug变得简单。我喜欢任何让生活变得简单的事情--难道你不是吗？

到目前位置，牢记Python的断言是一个调试的辅助工具，不是一套处理运行时错误的机制。
使用断言是为了让开发者能够更快发现可能导致bug的根源。
如果程序没有bug，断言错误永远不应该被主动抛出

让我们看一下可以用断言做的其它事情，后面我将会列出2个实践中常见的陷阱


### Python的断言语法
在使用一个Python的特性之前，最好学习一下它是怎样实现的。我们快速看看Python文档中断言的语法

    assert_stmt ::= "assert" expression1 ["," expression2]

在这个例子中，expression1是我们测试的条件，可选的expression2是一个断言失败后显示的错误信息。
在执行时，Python解释器吧每个断言转换成大致如下的语句:

    if __debug__:
        if not expression1:
        raise AssertionError(expression2)

这段代码里面有两件很有趣的事情:

在断言条件检查之前，对全局变量__debug__进行了额外的检查。__debug__是一个内建的bool值，
正常情况下这个值是true，如果需要优化，这个值是false
在后面常见陷阱部分，我们会对这个问题进行更深入的讨论

你还可以使用expression2传入一个可选的错误信息，它会在`AssertionError`的错误栈中显示。
这可以简化调试，比如下面的例子

    if cond == 'x':
     ... do_x()
     elif cond == 'y':
     ... do_y()
     else:
     ... assert False, (
     ... 'This should never happen, but it does '
     ... 'occasionally. We are currently trying to '
     ... 'figure out why. Email dbader if you '
     ... 'encounter this in the wild. Thanks!')
     
这难看吗？这确实是一种难看的写法，但在你遇到难以复现的bug的时候，这确实是一种有用的方法。


### Python中断言的常见陷阱

在继续之前，我还像指出两个和Python中断言相关的重要警告。
第一个和引入安全问题和bug有关，第二个和很容易写出无用断言的奇怪语法有关。

这听起来很恐怖，所以你最好至少遵守下面的两个警告

### 警告1：不要用断言进行数据验证

在使用Python断言时最大的警告：断言可以通过-0和--0命令行开关进行全局禁用，
CPython中的PYTHONOPTIMIZE环境变量也可以禁用断言

这会将所有的断言语句变成空操作：断言会在编译时被排除，并且永远不会被评估，这意味着没有任何断言会被执行

这是一个设计决策，在很多其它编程语言中也有类似的决策。
这种行为的副作用就是，使用断言对输入数据进行快速验证是极度危险的。

如果你的代码使用断言检查函数的参数是否包含一个"错误"的或者不在预期之内的值，
这可能会适得其反，导致一些bug或者安全漏洞.

让我们用一个简单例子说明这个问题. 假设你在使用Python搭建一个在线商城. 
在你应用的某个地方有一个函数根据用户的请求删除产品.

因为你刚刚学过断言，所以你很渴望在代码中使用断言，你写下了下面的代码.

    def delete_product(prod_id, user):
        assert user.is_admin(), 'Must be admin'
        assert store.has_product(prod_id), 'Unknown product'
        store.get_product(prod_id).delete()


仔细看看这个`delete_product`函数，如果断言被禁用，将会发生什么？

在这个三行代码的函数中，有两个很严重的问题，它们都是因为不正确地使用断言导致的.

- 使用断言检查管理员权限是很危险的. 如果在Python解释器中，断言被禁用，这将变成一个空操作.
所以任何人都可以删除产品.权限检查的代码甚至根本没有运行.这很有很能引入安全问题， 
  并且为攻击者摧毁或者严重损坏我们在线商城的数据打开大门.这不是一个好的用法.
  
- `has_product()`检查在断言禁用的时候会直接跳过.
  这意味着`get_product()`可以通过无效的产品id进行调用--这会导致更严重的bug.
  在最坏的情况下，这很可能是一个导致服务拒绝攻击的途径.
  比如，可能商城崩溃了，可能有人尝试删除了一个未知的商品，一个攻击者可能会通过无效的删除请求进行轰炸，进而导致服务中断.
  
我们怎么避免这些问题呢？答案就是永远不要使用断言进行数据验证.我们可以使用常规的if语句进行验证，在必要的时候抛出验证异常，比如:

    def delete_product(product_id, user):
        if not user.is_admin():
        raise AuthError('Must be admin to delete')
        if not store.has_product(product_id):
        raise ValueError('Unknown product id')
        store.get_product(product_id).delete()

这个更新后的例子还有一个好处: 
它现在抛出像ValueError或者AuthError(我们必须自己定义)这类语义上正确地异常，
而不是抛出一个不明确的断言异常.


### 警告2: 永不失败的断言

偶然写出永远为真的Python断言是一件相当容易的事情.我在过去就尝到过这样的苦果.
### TODO 补充一句

当你将元组做为第一个参数传给断言的时候，断言总是为真，从不失败.
下面的这个例子中的断言永远不会失败.

    assert(1 == 2, 'This should fail')

这和Python中不为空的元祖总是为真有关.
如果你传一个元祖给一个断言，会导致断言条件永远是真--进而导致上面的断言语无效，
因为断言语句永远不会失败，永远不会出发一个异常.

相对来说很容易因为直觉行为写出这种多行的糟糕断言。
比如，我很快乐的在我的一个测试套件中写了一堆无用的测试用例，这些用例给人以一种错误的安全感。
想象一下在你的一个单元测试中有如下的断言。

    assert (
     counter == 10,
     'It should have counted all the items'
     )
     
在第一次检测的时候，这个测试案例看起来没有任何问题。
然而它永远不会捕获错误的结果：无论变量counter的值为何，断言总是为真。
为什么会这样呢？因为它在对一个值为真的元组进行断言。

就像我说过的，这种做法很容易砸到自己的脚。一个很好的防止这种奇怪预发引起麻烦的对策是使用[code linter](https://dbader.org/blog/catching-bogus-python-asserts)
Python3的新版本会对这种模糊的断言给出警告。

顺便说一句，这就是为什么你应该始终对单元测试用例进行快速烟雾测试的原因。
确保它们确实可以失败然后继续写下一个。


### Python断言--总结

除了这些经过，我坚信Python的断言是一个没有被Python开发者充分使用的调试工具。

懂得断言怎么工作，在什么时候使用断言可以帮助你写出更容易维护的和更容易调试的代码。

这是一项很棒的学习技能，它可让你的Python知识更进一步，使你成为更全面的Pythonista。
我知道它节省了我数小时的调试工作。


### 重点
- Python的assert语句是一种调试辅助工具，可以使用内部自检的方式测试一个条件。
- 断言仅应用于帮助开发人员识别错误。它们不是用于处理运行时错误的机制。
- 可以通过解释器设置在全局禁用断言。


## 2.2 逗号放置

当你在Python中的列表，字典或集合中增删元素的时候，有一个有用的建议：以逗号作为所有行的结尾。

不确定我在说什么？让我来给出一个例子.想象一下你在代码里面有这样一个名字的列表.

    >>> names = ['Alice', 'Bob', 'Dilbert']

当你修改这个列表的时候，通过git的diff命令很难看到具体修改了什么。
很多的源码管理系统是基于行的，很难凸显在一行中的多出修改。

一个快速的解决方法是采用一种更新的代码风格，将列表，字典，集合像下面这样展开到多个行。

    >>> names = [
        ... 'Alice',
        ... 'Bob',
        ... 'Dilbert'
        ... ]

这样每一行就只有一个元素，添加，修改，删除一个元素在代码管理系统中很容易通过diff进行显示。
这是一个简单的改动，但是帮我避免了愚蠢的错误。这也让我团队的同事在review我的代码的时候更加轻松。

现在还有两个涉及到编辑的情况会引起困惑。当你在列表的末尾添加一个元素，或者你删除了最后一个元素，
你不得不修改逗号的位置以保持统一的代码风格。

比如你想要在列表添加另一个名字（_Jane_）。
如果你添加了*Jane*，你就需要修改在*Dilbert*这一行所在的逗号来避免错误。

    >>> names = [
        ... 'Alice',
        ... 'Bob',
        ... 'Dilbert' # <- Missing comma!
        ... 'Jane'
        ]
        

当你检查列表的内容时，你可能会感到惊奇：

    >>> names
        ['Alice', 'Bob', 'DilbertJane']

你看，Python把字符串*Dilbert*和*Jane*合并成了一个新的字符串`DilbertJane`。
这种所谓的“字符串字面串联”是有意设计的，在文档里也有说明。
这是一种音符难以捕获的bug到代码中，搬起石头砸自己的脚的匪夷所思的方式。

> 多个相邻的字符或者字节（使用空格分割）是允许的，
> 他们可能使用不同的引号，它们的意义和它们的连接体是一样的。

字符串字面连接在某些情况下还是一个有用的特性。比如，你可以用它减少分割长字符串到多个行时需要的反斜杆。

    my_str = ('This is a super long string constant '
               'spread out across multiple lines. '
               'And look, no backslash characters needed!')

另一方面，我们刚刚看到相同的功能如何快速导致一种问题。现在我们怎么解决这个问题？

在*Dilbert*行后面添加逗号会防止两个字符拼接成一个。
    
    >>> names = [
        ... 'Alice',
        ... 'Bob',
        ... 'Dilbert',
        ... 'Jane'
        ]

但是现在我们绕了一圈又回到了原点。为了添加一个新的名字到列表，我必须修改两行。
这导致在`git diff`时很难看到什么被修改了。是有人添加了一个新的名字，还是有人修改了*Dilbert*的名字？

幸运的是，Python的语法允许一些回旋余量来一劳永逸地解决这个逗号放置的问题。
你只需要训练自己采用一种在一开始就避免它的代码风格。让我具体来解释。

在Python中，你可以在列表，字典，集合中为每一个元素（包含最后一个元素）后面都添加一个逗号，
你只需要记得让每一行都使用逗号结尾，这样你就可以避免对是否放置逗号的判断。

最终的例子如下：
    
    >>> names = [
        ... 'Alice',
        ... 'Bob',
        ... 'Dilbert',
        ... ]

你有没有注意到Dilbert这一行最后的逗号，这样可以在新增或者删除元素的时候不必在更新逗号的位置。
这让你的代码保持一致，代码管理系统的diff操作更加清晰明了，代码评审的同事更加开心。
有时候，魔法藏在细节中，不是吗？

### 重点
- 合理的代码风格和逗号放置可以使你的列表，字典，或集合更容易维护
- Python 的字符串字面连体功能可以带来好处，也可能引入难以捕捉的错误。


## 2.3 上下文管理器和with语句

有时候Python中的`with`语句被当做一种晦涩的难懂的特性。
但是当你明白了其中的原理的时候，你会发现其中并没有什么魔法，
并且实际上`with`语句这个特性可以帮你写出整洁和易读的代码。

那么with语句有什么用呢？通过抽象资源的功能，
with语句有助于简化某些常见的资源管理模式并允许将它们分解和重用。

一个很好的观察这个功能的使用方法是查看Python标准库中的示例。
内置的`open()`函数为我们提供了一个很好的用例：

    with open('hello.txt', 'w') as f:
        f.write('hello, world!')

通常推荐使用with语句打开文件，
因为它可以确保程序在执行with语句的上下文后自动关闭打开的文件描述符。
最后，上面的代码示例可以转换为如下形式：

    f = open('hello.txt', 'w')
     try:
        f.write('hello, world')
     finally:
        f.close()
        
你可能会说这有点冗长。但是请注意`try ... finally`语句很重要。像下面这样写是不够的。
    
    f = open('hello.txt', 'w')
     f.write('hello, world')
     f.close()
     
如果在f.write()调用期间发生异常，这种实现将不能保证文件一定被关闭，因此我们的程序可能导致文件描述符泄漏。
这就是with语句如此有用的原因。它使获取和释放资源变得轻而易举。

另一个with语句被用于标准库的很好的例子是threading.Lock类：

    some_lock = threading.Lock()
     # Harmful:
     some_lock.acquire()
     try:
        # Do something...
     finally:
        some_lock.release()
     
     # Better:
     with some_lock:
        # Do something...


在这两种情况下，使用with语句可以抽象出大部分资源处理逻辑。
不必每次都写一个try... finally语句，可以使用with语句帮我们做这些事情。

with语句可以使处理系统资源的代码更具可读性。
通过自动清理或释放不再需要的资源，还可以帮助你避免错误或资源泄漏。

### 在你自己的对象中支持with语句

open()函数或threading.Lock类之所以可以使用with语句并不是因为任何特殊的地方或者魔法。
你可以通过实现所谓的上下文管理器来为自己的类和函数提供相同的功能。

什么是上下文管理器？这是一个对象要支持with语句需要遵从的简单“协议”。
如果你想让一个对象充当上下文管理器，你所需要做的就是在你的对象中添加__enter__和__exit__方法。
Python将在资源管理周期的适当时间调用这两个方法。

让我们看一下实际情况。一个简单实现的open()上下文管理器可能像下面这样：

    class ManagedFile:
        def __init__(self, name):
            self.name = name
         
        def __enter__(self):
            self.file = open(self.name, 'w')
            return self.file
         
        def __exit__(self, exc_type, exc_val, exc_tb):
            if self.file:
                self.file.close()

就像原始的open()示例做的，现在我们的ManagedFile类遵循了上下文管理器协议，支持了with语句：

    >>> with ManagedFile('hello.txt') as f:
        ... f.write('hello, world!')
        ... f.write('bye now')


当进入上下文时，Python调用`__enter__`获取资源。
当离开上下文时，Python调用`__exit__`释放资源。

编写基于类的上下文管理器并不是唯一的支持with语句的方式。
标准库中的`contextlib`模块提供了一些基本的上下文管理器协议之上的抽象。
如果你的使用场景与`contextlib`所提供的相匹配，使用它会更简单一点。

例如，你可以使用`contextlib.contextmanager`装饰器为资源定义基于生成器的工厂函数，该函数将自动支持with语句。
下面是我们使用这种技术重写的`ManagedFile`上下文管理器示例：

    from contextlib import contextmanager
    @contextmanager
    def managed_file(name):
        try:
            f = open(name, 'w')
            yield f
        finally:
            f.close()

    >>> with managed_file('hello.txt') as f:
    ... f.write('hello, world!')
    ... f.write('bye now')


在这种情况下，`managed_file()`是一个生成器，它首先获取资源，
之后，它会暂时中止自己的执行，在yield处产生资源，以便调用者可以使用它。
当调用者离开with上下文，生成器继续执行，以执行剩余的清理步骤，将资源释放。

基于类的实现和基于生成器的实现本质上是等效的。
你可能更喜欢其中的一个，具体取决于你认为哪种方法更具可读性。

基于@contextmanager的实现的缺点可能是它需要对Python的高级概念有一些了解，例如装饰器和生成器。
如果你需要获取相关的知识，请随意阅读本书中的相关章节。

再强调一次，在这里做出正确的选择取决于你和你团队的习惯以及你们对易读性的评价标准。

### 使用上下文管理器编写漂亮的接口

上下文管理器非常灵活，如果你创造性地使用with语句，则可以为模块和类定义便捷的API。
例如，如果我们要管理的“资源”是某种报告生成器程序中的文本缩进级别，我们可以这样编写代码：

    with Indenter() as indent:
        indent.print('hi!')
        with indent:
            indent.print('hello')
            with indent:
                indent.print('bonjour')
        indent.print('hey')

这几乎就像用于缩进文本的特定领域的语言（DSL）一样。
另外，请注意代码如何多次进入和离开同一个上下文管理器以更改缩进级别。
运行该代码段会得到以下输出并整齐地将格式化文本输出到控制台：

    hi!
        hello
            bonjour
    hey

那么，你将如何实现上下文管理器来支持此功能？

顺便说一句，这可能是一个很好的练习机会，可以让你准确了解上下文管理器如何工作。
因此，在你查看下面的实现之前，你最好将其作为一个练习，花一些时间尝试实现这个功能。

如果你准备看看我的实现，这里是一个使用基于类的上下文管理器实现此功能：

    class Indenter:
     def __init__(self):
     　  self.level = 0
    
    def __enter__(self):
    　   self.level += 1
    　   return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.level -= 1
        def print(self, text):
        print(' ' * self.level + text)


那还不错，不是吗？我希望你已经感觉到在Python程序中使用with语句和上下文管理器会更加优雅。
它们是一项出色的特性，可让你以更加符合Python风格的方式管理资源，并且让代码更易于维护。

如果你正在寻找另一个练习来加深理解，尝试实现一个使用time.time函数测量执行时间的上下文管理器。
一定要尝试同时编写基于装饰器和基于类的变体以理解两者之间的区别。

### 重点

- with语句通过将`try/finally`语句的标准用法封装在上下文管理器中以简化异常处理。
- 最常见的场景是用于安全地获取和释放系统资源。进入with语句时自动获取资源离开with语句时自动释放资源。
- 有效使用上下文管理器可以帮助你避免资源泄漏并且使你的代码更易于阅读。


## 2.4 下划线，双下划线和其他

单下划线和双下划线在Python变量和方法具有特殊的意义。
某些只是按照惯例向程序员提供一些提示，其中一些是Python解释器强制的。

如果你想知道，“Python变量和方法名称中的单下划线和双下划线有什么意思？”我会尽力给出答案。
在本章中，我们将讨论以下五个下划线模式和命名约定，以及它们如何影响Python程序的行为：

- 单下划线开头的变量： `_var`
- 单下划线结尾的变量：`var_`
- 双下划线开头的变量：`__var`
- 双下滑先开头和结尾的变量：`__var__`
- 单下划线变量： `_`

### 1. 单下划线开头的变量： `_var`

单下划线开头的变量和方法仅具有约定的含义。
这是对程序员的一种提示，这只是Python社区的一种约定，但这不会影响程序的行为。

单下划线前缀是用来提示这个变量或方法供内部使用。
PEP 8（最常用的Python代码风格指南）中定义了这个约定。

但是，Python解释器没有强制执行这个约定。
Python在“private”和“public”之间没有像Java一样明显的区别。
在变量和方法名称前面添加单下划线更像是一种警告：“这并不是要成为此类的公共接口，最好不要直接调用它。”

看下面的例子：

    `class Test:
     def __init__(self):
         self.foo = 11
         self._bar = 23
     
如果你实例化该类并尝试访问该类在其__init__方法中定义的foo和_bar属性时会发生什么？

让我们试试：
    
    >>> t = Test()
    >>> t.foo
    11
    >>> t._bar
    23
     
变量`_bar`中的前导下划线并没有阻止我们“进入”该类并获得它的值。

那是因为Python中的单下划线前缀只是一个约定俗成的惯例——至少在变量和方法名称中是如此。
但是，从模块导入时下划线会影响导入的行为。
假设你在一个名为my_module的模块中有如下代码：

    # my_module.py:
    def external_func():
       return 23
    
    def _internal_func():
       return 42

现在，如果你使用通配符来从模块导入，Python不会导入带有下划线的名称（除非模块定义了__all__列表来覆盖此行为）：

    >>> from my_module import *
    >>> external_func()
    23
    >>> _internal_func()
    NameError: "name '_internal_func' is not defined"
     
顺便说一句，应该避使用免通配符导入，因为它们导致名称空间混乱。
为了清楚起见，最好坚持使用常规导入。
与通配符导入不同，常规导入不受前导下划线命名的影响：

    >>> import my_module
    >>> my_module.external_func()
    23
    >>> my_module._internal_func()
    42
     
我知道这可能有点令人困惑。
如果你坚持PEP 8应该避免通配符导入的原则，那么你只需要记住：

单个下划线是Python的命名约定，表示名称仅供内部使用。
它通常不由Python解释器强制校验，仅仅是对程序员的提示。


### 2.单下划线结尾： `var_`

有时，某个变量最合适的名称已经被Python中的关键字占用。
因此，class或def之类的名称不能在Python中用作变量名。
在这种情况下，你可以通过附加单个下划线解决命名冲突：

    >>> def make_object(name, class):
    SyntaxError: "invalid syntax"
    >>> def make_object(name, class_):
    ... pass

总之，使用单个下划线结尾来避免与Python关键字的命名冲突是一个约定，PEP 8中定义和解释了这个约定。

### 3.双下划线开头： `__var`

到目前为止，我们讨论的命名规范都是一些约定俗成。
使用以双下划线开头的Python类属性（变量和方法）时，事情就不一样了。

Python解释器会重写双下划线前缀的属性名称以避免子类中的命名冲突。

这也称为名称处理——当类在被扩展的时候解释器通过一种更不易出现名称冲突的方法更改变量名称。

我知道这听起来很抽象。所以我放了这一小段代码来做实验：

    class Test:
        def __init__(self):
            self.foo = 11
            self._bar = 23
            self.__baz = 23

我们使用内建的dir()方法看看这个对象的属性：

    >>> t = Test()
    >>> dir(t)
    ['_Test__baz', '__class__', '__delattr__', '__dict__',
    '__dir__', '__doc__', '__eq__', '__format__', '__ge__',
    '__getattribute__', '__gt__', '__hash__', '__init__',
    '__le__', '__lt__', '__module__', '__ne__', '__new__',
    '__reduce__', '__reduce_ex__', '__repr__',
    '__setattr__', '__sizeof__', '__str__',
    '__subclasshook__', '__weakref__', '_bar', 'foo']

这为我们提供了包含对象属性的列表。
让我们来看看这个清单找一下我们的原始变量名称`foo`，`_bar`和`__baz`。
我保证你会注意到一些有趣的变化。

首先，`self.foo`变量在属性列表中还是`foo`，没有什么变化。
接下来，`self._bar`也一样，它显示为`_bar`。
就像我之前说过的，下划线开头只是一个约定，是一钟对程序员的提示。

但是，对于`self.__baz`来说，情况看起来有些不同。当你在该列表中搜索`__baz`，你会看到没有变量有这个名字。
    
对于`__baz`来说，发生了什么？

如果仔细观察，会发现这个对象有一个名为_Test__baz的属性。
这就是Python解释器的名称转换。这样做是为了防止变量被子类覆盖。

让我们创建另一个扩展`Test`类的类，并尝试覆盖其在初始化函数中添加的现有属性：

    class ExtendedTest(Test):
        def __init__(self):
            super().__init__()
            self.foo = 'overridden'
            self._bar = 'overridden'
            self.__baz = 'overridden'

现在，你认为在这个ExtendedTest类的实例中，`foo`，`_bar`和`__baz`的值会是什么？让我们来看看：

    >>> t2 = ExtendedTest()
    >>> t2.foo
    'overridden'
    >>> t2._bar
    'overridden'
    >>> t2.__baz
    AttributeError:
    "'ExtendedTest' object has no attribute '__baz'"

等等，为什么我们尝试获取`t2.__baz`的值时会抛出`AttributeError`？
名称被改变了！该对象甚至没有`__baz`属性：

    >>> dir(t2)
    ['_ExtendedTest__baz', '_Test__baz', '__class__',
    '__delattr__', '__dict__', '__dir__', '__doc__',
    '__eq__', '__format__', '__ge__', '__getattribute__',
    '__gt__', '__hash__', '__init__', '__le__', '__lt__',
    '__module__', '__ne__', '__new__', '__reduce__',
    '__reduce_ex__', '__repr__', '__setattr__',
    '__sizeof__', '__str__', '__subclasshook__',
    '__weakref__', '_bar', 'foo', 'get_vars']

你看，`__baz`变成了`_ExtendedTest__baz`以防止被意外修改。但是原来的`_Test__baz`仍然存在：

    >>> t2._ExtendedTest__baz
    'overridden'
    >>> t2._Test__baz
    42

双下划线名称改写对程序员是完全透明的。下面的示例将确认这一点：

    class ManglingTest:
        def __init__(self):
            self.__mangled = 'hello'

        def get_mangled(self):
            return self.__mangled
    >>> ManglingTest().get_mangled()
    'hello'
    >>> ManglingTest().__mangled
    AttributeError:
    "'ManglingTest' object has no attribute '__mangled'"

名称转换也适用于方法名称吗？当然！名称转换会影响类上下文中以两个下划线开头的所有名称：

    class MangledMethod:
        def __method(self):
            return 42
        
        def call_it(self):
            return self.__method()
    >>> MangledMethod().__method()
    AttributeError:
    "'MangledMethod' object has no attribute '__method'"
    >>> MangledMethod().call_it()
    42

这是名称转换行为的另一个示例，也许令人惊讶：

    _MangledGlobal__mangled = 23
    class MangledGlobal:
        def test(self):
            return __mangled

    >>> MangledGlobal().test()
    23

在此示例中，我将`_MangledGlobal__mangled`声明为全局变量。
然后我在`MangledGlobal`类的上下文中访问了这个变量。
由于名字改变，我得以在类的`test()`方法内使用`__mangled`引用`_MangledGlobal__mangled`全局变量。

因为`__mangled`以两个下划线开头，Python解释器会自动将`__mangled`转换到`_MangledGlobal__mangled`。
这表明名称转换不仅仅作用在类属性。它适用于在类上下文中以两个下划线字符开头的任何名称。

这要消化很多东西。

老实说，我没有不假思索就写下这些例子。我花了一些时间研究才做到这一点。
我已经使用Python多年了，但是像这样的规则和特殊情况我也做不到信手拈来。

有时，对于程序员而言，最重要的技能是“模式识别”，知道在哪里查找相关的答案。
如果你此时有一点感觉不堪重负，别担心。
花时间把玩一下本章中的一些示例。

让这些概念足够深入，你才能够认识到名称转换的理念以及我展示的一些其他行为。
如果有一天你猛然遇到它们，你会知道在文档中寻找什么。

#### 什么是dunders

如果你听说过一些经验丰富的Python使用者谈论Python或观看过一些会议演讲，
你可能已经听说过dunder一词。如果你想知道这是什么，那么，答案在这里：

双下划线在Python中通常称为“dunder”。
原因是双下划线出现的频率很高，为了更容易发音，
Python开发者通常将“dunder”作为“double underscore”的缩写。

例如，`__baz`可以发音为“dunder baz”。
同样地，`__init__`可以发音为“dunder init”，即使有人可能认为它应该是“dunder init dunder”。
但这只是另一个命名约定。就像是Python开发人员的秘密握手。

### 4.双下划线开头和结尾： `__var__`

也许令人惊讶的是，如果名称以双下划线开头并以双下划线结尾，则不会进行名称转换。
Python解释器没有对双下划线开头和双下划线结尾的变量进行任何修改：

    class PrefixPostfixTest:
        def __init__(self):
        self.__bam__ = 42
    >>> PrefixPostfixTest().__bam__
    42

但是，双下划线开头和双下划线结尾的名称被语言保留以供特殊使用。
像`__init__`方法、`__call__`方法都符合这个规则。

这些双下划线开头和结尾的方法通常被称为魔术方法，但是在Python社区中，包括我自己在内的的许多人都不喜欢这个词。
这个词意味着不鼓励使用双下划线开头和结尾的方法，但是事实完全不是这样。
它们是Python的一个核心特性，应该根据需要加以使用。他们并没有什么“魔法”。

但是，就命名约定而言，在自己的代码中最好不要使用以双下划线开头和结尾的名称，以避免与将来与Python语言的更改产生冲突。

### 5.单下划线： `_`

按照惯例，有时使用单个独立的下划线表示变量是临时的或无关紧要的。
例如，在以下循环中，我们不需要访问正在运行的索引，我们可以使用“_”表示它只是一个临时值：

    >>> for _ in range(32):
    ... print('Hello, World.')

你还可以在解包语句中使用单个下划线来忽略无关紧要的特定值。
同样，这也仅仅是一个惯例，并不会触发Python解释器的任何特殊行为。
单下划线是一个有效变量名，只是有时候被这样用了而已。

在下面的代码示例中，我将一个元组解包为单独的变量，我只对颜色和里程字段感兴趣。 
但是，为了使解包表达式能够正常工作，我需要将元组中包含的所有值分配给变量。
这就是占位符变量“_”的作用：

    >>> car = ('red', 'auto', 12, 3812.4)
    >>> color, _, _, mileage = car
    >>> color
    'red'
    >>> mileage
    3812.4
    >>> _
    12

除了用作临时变量外，“_”在大多数情况下是表示解释器执行的最后一个表达式结果的特殊变量。
如果你正在解释器的会话中工作并且想访问先前计算的结果，这会很方便：

    >>> 20 + 3
    23
    >>> _
    23
    >>> print(_)
    23

如果你正在动态地构建对象并且想要在不给他们分配名称的情况下与其进行交互，这也很方便：

    >>> list()
    []
    >>> _.append(1)
    >>> _.append(2)
    >>> _.append(3)
    >>> _
    [1, 2, 3]

###　重点

- 单个前导下划线“`_var`”：命名惯例，表示名称仅供内部使用。通常不由Python解释器强制执行（通配符导入除外）只是对程序员的提示。
- 单尾划线“`var_`”：按惯例用于避免与Python关键字冲突。
- 双引号下划线“`__var`”：在类上下文中使用时会触发名称转换，由Python解释器强制执行。
- 前后双下划线“`__var__`”：表示由Python语言定义的特殊方法。避免自己的属性使用这种命名方案。
- 单个下划线“`_`”：有时用作临时或无关紧要的变量的名称。此外，它还表示在Python REPL会话中发送最后一个表达式的结果。


## 2.5 关于字符串格式化的真相

还记着在Python之禅有一条“做某件事应该有一种明显的方式？”
当你发现在Python中有四种主要的方法可以进行字符串格式化时，你可能会抓狂。

在本章中，我将演示这四种字符串格式化方法的使用方法以及它们各自的优缺点。
我还将给出我如何选择最佳字符串格式化方法的“经验法则”。

因为我们有很多内容要讲，所以我们直接开始。
为了有一个简单的实验示例，假设我们有以下变量可以使用：

    >>> errno = 50159747054
    >>> name = 'Bob'
     
根据这些变量，我们想输出带有以下错误消息字符串：

    'Hey Bob, there is a 0xbadc0ffee error!'
    
现在，该错误可能真的浪费了开发人员星期一的早上！
但是我们今天是在这里讨论字符串格式。因此，让我们开始工作。

### #1 – 老式的字符串格式化

Python中的字符串具有唯一的内置操作标识符，可以使用％号访问。
这是实现简单的位置格式化的捷径。
如果你曾经使用过C语言中的printf函数，你会立即意识到它是如何工作的。
这是一个简单的示例：

    >>> 'Hello, %s' % name
    'Hello, Bob'
    
我在这里使用％s来告诉Python在哪里将其替换成以字符串表示的值。
这就是所谓的“老式”字符串格式化。

在“老式”字符串格式化中，还有其他格式说明符可用于控制输出的字符串。
例如，可以将数字转换为十六进制表示法或添加空格以生成格式正确的表格和报告。

在这里，我使用％x格式说明符将int值转换为十六进制数字并用字符串表示出来：

    >>> '%x' % errno
    'badc0ffee'
    
如果你想在单个字符串中进行多个替换，“老式”字符串格式语法会稍有变化。
因为％操作符只需要一个参数，你需要将右侧的参数包裹在一个元组中，像下面这样：

    >>> 'Hey %s, there is a 0x%x error!' % (name, errno)
    'Hey Bob, there is a 0xbadc0ffee error!'
     
如果你将一个字典传递给％操作符，你还可以在字符串格式化中使用变量名指定变量：

    >>> 'Hey %(name)s, there is a 0x%(errno)x error!' % {
    ... "name": name, "errno": errno }
    'Hey Bob, there is a 0xbadc0ffee error!'
     
这会使字符串格式化更易于维护和修改。
你不用担心传递的值顺序与字符串格式化中引用顺序的匹配问题。
当然，缺点是需要打更多的字。

我敢肯定你一直想知道为什么这种printf风格的格式化被打上“老式”的标签。 
老式的字符串格式化在技术上已被我们将要谈到的“新式”格式化取代。 
但是，虽然不再被强调，“老式”格式化并未被弃用。
最新版本的Python仍支持该功能。


### #2 新式字符串格式化

Python 3引入了一种新的字符串格式化方式，后来也支持了Python 2.7。
这种“新式”字符串格式化弃用了％操作符，让字符串格式化更规整。
现在可以通过在字符串对象上调用format()函数来进行格式化。

你可以使用format()函数进行简单的基于位置的格式化，这和使用“老式”格式化很像：

    >>> 'Hello, {}'.format(name)
    'Hello, Bob'

你在按名称进行变量替换时，可以按照任何你喜欢的顺序进行。
这是一个非常强大的功能，因为它允许重新排列显示顺序而不用更改传递给format函数的参数：

    >>> 'Hey {name}, there is a 0x{errno:x} error!'.format(
    ... name=name, errno=errno)
    'Hey Bob, there is a 0xbadc0ffee error!'

将int变量格式化为十六进制字符串的语法也有变化。
现在通过在变量名后添加“：x”后缀，我们添加一个了格式化规范。

总体而言，新式字符串格式化在没有复杂化简单用例的前提下，让字符格式化语法变得更加强大。
Python文档中关于[字符串格式化](https://docs.python.org/3/library/string.html#string-formatting)
的部分值得一读。

在Python 3中，这种“新式”字符串格式化更受青睐。
但是，从Python 3.6开始，还有一中更好的字符串格式化方法。在下一节中，我将告诉你所有的相关信息。

### #3 文字字符串内插(Python 3.6+)

Python 3.6添加了另一种格式化字符串的方式，称为格式化的文字字符串。
这种格式化字符串的新方法使你可以在字符串常量中嵌入Python表达式。
这是一个简单的示例，可以让你对该功能有所了解：

    >>> f'Hello, {name}!'
    'Hello, Bob!'

这种新的格式语法功能强大。由于你可以嵌入任意Python表达式，因此甚至可以进行运算，
比如：

    >>> a = 5
    >>> b = 10
    >>> f'Five plus ten is {a + b} and not {2 * (a + b)}.'
    'Five plus ten is 15 and not 30.'

格式化的字符串文字依赖Python解析器的一个特性，这个特性可以将f字符串转换为一系列字符串常量和表达式。
然后，他们会被拼接起来构建最终的字符串。

假设我们有一个包含f字符串的greet()函数：

    >>> def greet(name, question):
    ...     return f"Hello, {name}! How's it {question}?"
    ...
    >>> greet('Bob', 'going')
    "Hello, Bob! How's it going?"

当我们反汇编该函数并对其进行检查时，我们可以看到该函数中的f字符串已经被转换为类似于下面的东西：

    >>> def greet(name, question):
    ...     return ("Hello, " + name + "! How's it " +
                    question + "?")

实际的实现要比上面的实现快一点，因为它使用了[BUILD_STRING](https://bugs.python.org/issue27078)
进行优化。
但从功能上讲，它们是相同的：

    >>> import dis
    >>> dis.dis(greet)
        2     0 LOAD_CONST 1 ('Hello, ')
              2 LOAD_FAST 0 (name)
              4 FORMAT_VALUE 0
              6 LOAD_CONST 2 ("! How's it ")
              8 LOAD_FAST 1 (question)
              10 FORMAT_VALUE 0
              12 LOAD_CONST 3 ('?')
              14 BUILD_STRING 5
              16 RETURN_VALUE

文字字符串也支持现有的str.format()语法。
这样你就可以解决我们在前两节中讨论的格式化问题：

    >>> f"Hey {name}, there's a {errno:#x} error!"
    "Hey Bob, there's a 0xbadc0ffee error!"

Python的新格式化字符串文字类似于JavaScript在ES2015中添加的文字模板。
我认为它们是Python很好的一个补充，并且我已经在日常的Python 3工作中开始使用它们。
你可以在Python官方文档中了解有关格式化字符串的更多信息。

### #4-字符串模板

在Python中用于字符串格式化的另一种技术是字符串Template。
这是一种更简单，功能没有那么强大的机制，但在某些情况下，这可能正是你所寻找的特性。

让我们看一个简单的用于问候示例：

    >>> from string import Template
    >>> t = Template('Hey, $name!')
    >>> t.substitute(name=name)
    'Hey, Bob!'
     
在这里我们需要从Python的内置模块导入字符串模版。
模板字符串不是语言的核心功能，但由标准库中的一个模块提供。

另一个区别是模板字符串不允许格式标识符。
因此，为了使错误显示字符串正常输出，我们需要自己将整形错误号转换为十六进制字符串：

    >>> templ_string = 'Hey $name, there is a $error error!'
    >>> Template(templ_string).substitute(
    ... name=name, error=hex(errno))
    'Hey Bob, there is a 0xbadc0ffee error!'
     
效果很好，你可能想知道应该在什么时候使用模板字符串。
我认为，模板字符串最好的使用场景是处理用户生成的格式字符串。
由于降低了复杂度，模板字符串是一个更安全的选择

其他更复杂的字符串格式化技术可能会在你的程序中引入安全漏洞。
例如，格式字符串可以访问程序中的任意变量。

这意味着，如果恶意用户可以提供格式字符串，他们就有可能泄露密钥和其他敏感信息！
这是关于如何使用这种攻击的一个简单例子：

    >>> SECRET = 'this-is-a-secret'
    >>> class Error:
    ...     def __init__(self):
    ...         pass
    >>> err = Error()
    >>> user_input = '{error.__init__.__globals__[SECRET]}'
    # Uh-oh...
    >>> user_input.format(error=err)
    'this-is-a-secret'
     
了解潜在的攻击者是怎么从格式字符串访问__globals__字典来获取我们的SECRET字符串了吗？
吓死了，呵呵！模板字符串阻断了这种攻击方式，所以如果你要将用户输入进行字符格式化，字符串模板是一个更安全的选择：

    >>> user_input = '${error.__init__.__globals__[SECRET]}'
    >>> Template(user_input).substitute(error=err)
    ValueError:
    "Invalid placeholder in string: line 1, col 1"
     
### 我应该使用哪种字符串格式化方法？

我完全明白，在Python中有这么多格式化字符串的方式可供选择会让人感到非常混乱。
这时最好列出一些流程图...

但是我不会那样做。相反地，我会尝试将其归结为我在编写Python代码时遵循的一些简单经验法则。

你可以在不清楚选择那种字符串格式化方法的时候，根据具体情况使用这些经验。

### 笔者的Python字符串格式化经验

如果你的格式字符串是用户提供的，请使用字符串模板以避免出现安全问题。
如果你使用的是Python 3.6及更高版本，请使用字符串内插，如果不是，使用“新式”字符串格式化。

### 重点

 - Python有不止一种格式化字符串的方式。
 - 每种方法都有其各自的优缺点。应该使用哪种方法应该根据你的使用场景确定。
 - 如果你在确定使用哪种字符串格式化方法时遇到麻烦，请尝试我的“字符串格式化经验法则”

## 2.6 Python之禅

我知道，就Python书籍而言，有下面的内容是很常见的。
但Tim Peters的Python之禅真的无法绕过。
多年以来通过时常重新审视Python之禅，让我受益匪浅，Tim的话使我成为更好的程序员，希望你也一样。

此外，Python Zen确实很重要，因为它包含在语言的复活节彩蛋中。
只需在Python解释器输入并运行以下命令：

    >>> import this

### Python之禅

优美胜于丑陋

明了胜于晦涩

简洁胜于复杂

复杂胜于凌乱

扁平胜于嵌套

间隔胜于紧凑

可读性很重要

即便假借特例的实用性之名，也不可违背这些规则

不要包容所有错误，除非你确定需要这样做

当存在多种可能，不要尝试去猜测

而是尽量找一种，最好是唯一一种明显的解决方案

虽然这并不容易，因为你不是 Python 之父

做也许好过不做，但不假思索就动手还不如不做

如果你无法向人描述你的方案，那肯定不是一个好方案

如果你很容易向人描述你的方案，那可能是一个好方案

命名空间是一种绝妙的理念，我们应当多加利用

